### PR TITLE
ARM64: Fix Frame with compLocallocUsed

### PR DESCRIPTION
**Pull request by kyulee1**
When compLocallocUsed is true, stack pointer is changed during the run.
So, JIT should restore stack pointer from frame pointer in the epilog.
Previously, we simpliy copied fp to sp, which is only valid for frameType
= 1.
This implements a right restoring process for other frameTypes (2 and 3).
For frameType = 3, I also simplified creating prolog sequence to support
this feature consistently.

### CODE REVIEW

**Comment by kyulee1** [comment]
@dotnet-bot test Windows_NT arm64

File: src/jit/codegencommon.cpp [hunk 1]
```diff
@@ -9406,9 +9422,6 @@ void                CodeGen::genFnEpilog(BasicBlock* block)
             compiler->unwindBegEpilog();
             unwindStarted = true;
         }
-        // mov FP into SP
-        inst_RV_RV(INS_mov, REG_SPBASE, REG_FPBASE);
-        compiler->unwindSetFrameReg(REG_FPBASE, 0);
```
Comment (kyulee1) [inline src/jit/codegencommon.cpp]: I don't think this unwind code is right, so I didn't copy it for frameType =1 case -- we are not setting fp from sp but sp from fp.
Likewise, we'd like to adjust sp from fp (for other frameTypes), but I'm not sure how we express such things in unwind code. It appears sp adjustment api is only for delta from the current sp. Or don't we care at all in epilog? @BruceForstall Any comment?

**Comment by kyulee1** [comment]
@dotnet/jit-contrib PTAL

File: src/jit/codegencommon.cpp (hunk 1 above)
Comment (BruceForstall) [inline src/jit/codegencommon.cpp]: The epilog unwind codes are the inverse of the prolog unwind codes: you kind of need to think of them backwards. Ideally, the prolog and epilog codes are exact inverses, which allows for a smaller unwind codes encoding.

---
In reply to: [57753045](https://github.com/dotnet/coreclr/pull/3966#discussion_r57753045) [](ancestors = 57753045)

File: src/jit/codegencommon.cpp [hunk 2]
```diff
@@ -6343,13 +6329,27 @@ void            CodeGen::genPopCalleeSavedRegistersAndFreeLclFrame(bool
         if ((compiler->lvaOutgoingArgSpaceSize == 0) && (totalFrameSize < 512))
         {
             frameType = 1;
+            if (compiler->compLocallocUsed)
+            {
+                // Restore sp from fp
+                //      mov sp, fp
+                inst_RV_RV(INS_mov, REG_SPBASE, REG_FPBASE);
+            }
```
Comment (BruceForstall) [inline src/jit/codegencommon.cpp]: Every single instruction in the epilog needs an unwind code. You don't have any here, or below, in your new code.

**Comment by kyulee1** [comment]
The failures for other archs (which are unrelated) are from Jenkin's reboot, which I won't test again.
Arm64 failures (NonVirtualCall/OverwriteArray/b431011) are in the baseline as well -- we should update Tests.lst to reflect this. _il_relrotarg_float/_il_dbgrotarg_float failing are unstable -- the local run pass.

File: src/jit/codegencommon.cpp (hunk 2 above)
Comment (kyulee1) [inline src/jit/codegencommon.cpp]: This is the question that I'm asking you in other code-snippet.

File: src/jit/codegencommon.cpp (hunk 1 above)
Comment (kyulee1) [inline src/jit/codegencommon.cpp]: Can you provide an example of how I should set up unwind code for this case -- we restore sp from fp?

**Comment by BruceForstall** [comment]
As you make these fixes, please update the "ARM64 JIT frame layout" document (currently not on GitHub) with any additional details or fixes that are required.

File: src/jit/codegencommon.cpp [hunk 3]
```diff
@@ -5679,25 +5679,11 @@ void CodeGen::genPushCalleeSavedRegisters()
             assert(spAdjustment3 > 0);
             assert((spAdjustment3 % 16) == 0);
 
-            // Try to push the frame pointer setup down, so the unwind codes match better (there is no corresponding instruction in the epilog).
-            bool isFPEstablished = false;
-            int maxFPOffset = spAdjustment3 + alignmentAdjustment2;
-            if (!emitter::emitIns_valid_imm_for_add(maxFPOffset, EA_PTRSIZE))
-            {
-                getEmitter()->emitIns_R_R_I(INS_add, EA_PTRSIZE, REG_FPBASE, REG_SPBASE, alignmentAdjustment2);
```
Comment (BruceForstall) [inline src/jit/codegencommon.cpp]: Make sure you add or adjust the comment above related to frameType==3 to reflect the reality of the frame setup/restore changes you've made.

File: src/jit/codegencommon.cpp (hunk 1 above)
Comment (BruceForstall) [inline src/jit/codegencommon.cpp]: If you're going to remove the instruction, you can remove this whole clause. There's no need to call unwindBegEpilog() if we're not generating any unwind codes (or instructions) here.

---
In reply to: [57769409](https://github.com/dotnet/coreclr/pull/3966#discussion_r57769409) [](ancestors = 57769409,57753045)

Comment (BruceForstall) [inline src/jit/codegencommon.cpp]: This is exactly the case. The 0 is the offset, if you use an add/sub to establish the fp.

---
In reply to: [57773432](https://github.com/dotnet/coreclr/pull/3966#discussion_r57773432) [](ancestors = 57773432)

**Comment by kyulee1** [comment]
Fix is made per feedbacks. And also did more refactoring about unwindBegEpilog.

**Comment by BruceForstall** [comment]
LGTM

**Comment by briansull** [comment]
LGTM

**Pull request merged**